File: shopping-ai-backend/schemas/product.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class SearchRequest(BaseModel):
    session_id: int
    keywords: list[str] = Field(default_factory=list)
    page: int = 1
    page_size: int = 20


class ProductCard(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int | None = None
    platform: str
    external_product_id: str
    title: str
    image_url: str
    price: Decimal
    original_price: Decimal | None = None
    shop_name: str | None = None
    shop_type: str | None = None
    sales_count: int | None = None
    coupon_info: str | None = None
    product_url: str
    attributes: dict | None = None
    score: Decimal | None = None
    recommend_reason: str | None = None


class SearchResponse(BaseModel):
    session_id: int
    products: list[ProductCard]


class ProductDetail(ProductCard):
    created_at: datetime | None = None
    updated_at: datetime | None = None


class PricePoint(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    recorded_at: datetime
    price: Decimal


class PriceHistoryResponse(BaseModel):
    product_id: int
    currency: str = "CNY"
    points: list[PricePoint]


class FavoriteRequest(BaseModel):
    product_id: int
    device_id: str | None = None


class FavoriteResponse(BaseModel):
    favorite_id: int
    product_id: int
    created_at: datetime

File: shopping-ai-app/types/product.ts
export type RecognizedItem = {
  category: string;
  brand?: string | null;
  color?: string | null;
  style?: string | null;
  material?: string | null;
  keywords: string[];
  confidence: number;
  raw_result?: Record<string, unknown> | null;
  provider: "gemini" | "mock" | string;
  is_mock: boolean;
  fallback_reason?: string | null;
  raw_model_output?: string | null;
};

export type Product = {
  id: number;
  platform: "jd" | "pdd" | "mock" | string;
  external_product_id: string;
  title: string;
  image_url: string;
  price: number;
  original_price?: number | null;
  shop_name?: string | null;
  shop_type?: "official" | "authorized" | "normal" | string | null;
  sales_count?: number | null;
  coupon_info?: string | null;
  product_url: string;
  attributes?: Record<string, unknown> | null;
  score?: number | null;
  recommend_reason?: string | null;
};

export type FilterCondition = {
  min_price?: number | null;
  max_price?: number | null;
  platforms: string[];
  shop_type?: string | null;
  sort_by?: string | null;
  keywords_include: string[];
  keywords_exclude: string[];
};

export type PricePoint = {
  recorded_at: string;
  price: number;
};

export type PriceHistory = {
  product_id: number;
  currency: string;
  points: PricePoint[];
};

export type HistoryItem = {
  session_id: number;
  image_url: string;
  recognized_summary?: string | null;
  product_count: number;
  created_at: string;
};

export type UploadResult = {
  upload_id: string;
  session_id: number;
  image_url: string;
};

export type ApiEnvelope<T> = {
  data: T;
};
